pkg/util/dynamichelper/dynamichelper_test.go

```diff
@@ -645,3 +653,89 @@ func TestMakeURLSegments(t *testing.T) {
 		})
 	}
 }
+
+func TestEnsure(t *testing.T) {
+	controller := gomock.NewController(t)
+	defer controller.Finish()
+	manager := mock_dynamichelper.NewMockGVRInterface(controller)
+
+	manager.EXPECT().Resolve(
+		gomock.Eq("Event"),
+		gomock.Any(),
+	).DoAndReturn(func(groupKind, optionalVersion string) (*schema.GroupVersionResource, error) {
+		return &schema.GroupVersionResource{
+			Group:    "",
+			Version:  "v1",
+			Resource: "event",
+		}, nil
+	})
+
+	h := http.Header{}
+	h.Add("Content-Type", "application/yaml")
+	var d kruntime.Object = &corev1.Event{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Event",
+			APIVersion: "v1",
+		},
+	}
+	// Prepare the encoder
+	serializer := kjson.NewSerializerWithOptions(
+		kjson.DefaultMetaFactory, scheme.Scheme, scheme.Scheme,
+		kjson.SerializerOptions{Yaml: true},
+	)
+	yaml := scheme.Codecs.CodecForVersions(serializer, nil, schema.GroupVersions(scheme.Scheme.PrioritizedVersionsAllGroups()), nil)
+
+	buf := new(bytes.Buffer)
+	err := yaml.Encode(d, buf)
+	if err != nil {
+		t.Errorf("Failed to encode document: %s", err.Error())
+	}
+
+	frc := &fake.RESTClient{
+		GroupVersion:         schema.GroupVersion{Group: "", Version: "v1"},
+		NegotiatedSerializer: scheme.Codecs.WithoutConversion(),
+		Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
+			switch req.Method {
+			case "GET":
+
+				switch req.URL.Path {
+				case "/api/v1/event":
+					return &http.Response{
+						StatusCode: http.StatusOK,
+						Header:     h,
+						Body:       io.NopCloser(buf),
+					}, nil
+				}
```

Comment:
LGTM, may be just add some more test cases (e.g GET request to another URL returns error)
```
case "/api/v1/other-url":
	// New test case: GET request to another URL returns 404 status code
	return &http.Response{
		StatusCode: http.StatusNotFound,
		Header:     h,
		Body:       io.NopCloser(bytes.NewBufferString("Not found")),
	}, nil
}
```

Reply:
Great catch! The reason I'm not testing the not found case is that the Ensure function is designed to create the object, if it doesn't already exist. So if the object can't be found it takes the "create this document" code path - this is an existing codepath that's in production and currently works. I agree that it should be tested, however... At this point in the PR to make this test one would have to mock the restcli or replace the restcli call (the preferred method) which would involve adding the ability to create objects to the GVR resolver. This functionality is actually added by Jeff's PR as I remember, so my intention here is not to duplicate work, or to change existing code that is already in production.